### ISSUE TITLE
Avoid ClassCastException when comparing two column values without natural order in Record::compareTo

### ISSUE THREAD

**Issue by amitgol** [jOOQ/jOOQ#10614]
### Expected behavior and actual behavior:
Given 2 equal records, it is expected that when `record1.equals(record2)` returns true then also `record1.compareTo(record2)` will return 0. Instead a ClassCastException is thrown. This happens when one of the fields is of a type that is not Comparable, for example a JsonNode when using type binding.

This is a regression when upgrading from jooq 3.10.x to 3.12.x 

Issue becomes a pain point when using groovy, for example for spock framework, since groovy interprets `a == b` as `a.compareTo(b) == 0` when a and b are instances of `Comparable ` (See http://docs.groovy-lang.org/docs/latest/html/documentation/#_identity_operator)

### Steps to reproduce the problem

Generate records with binding, for example with PostgresJSONGsonBinding as explained in the [documentation](https://www.jooq.org/doc/3.12/manual/code-generation/custom-data-type-bindings/).

Get 2 identical records, for example by fetching by same Id:

```
def record1 = dslContext.selectFrom(MY_TABLE).where(ID.eq(1)).fetchOne()
def record2 = dslContext.selectFrom(MY_TABLE).where(ID.eq(1)).fetchOne()

assert record1.compareTo(record2) == 0
```
Expected that assertions will not fail

### Versions:

- jOOQ: 3.12.4
- Java: 8
- Database: Postgres 10
- OS: MacOS
- JDBC Driver: 42.2.5

**Comment by lukaseder** [comment]
Thank you very much for your report. There are 2 things to consider here:

1. What caused the regression, or rather, why did this work before, for you? Do you still have the old behaviour available, perhaps? Was there simply no `ClassCastException`, but undefined behaviour instead (e.g. all `JsonNode` instances comparing as equal)?
2. What should be the behaviour, because we cannot decide how to compare `JsonNode` for you

Regarding 2), I think a reasonable solution to this would be a `ComparatorProvider` SPI that allows for injecting `Comparator<T>` instances for `Class<T>` references:

```java
public interface ComparatorProvider {
    <T> Comparator<T> comparator(Class<T> klass);
}
```

Since records have a `Configuration` attached, by default, they could use this SPI.

**Comment by amitgol** [comment]
Thank you Lukas for the prompt response!

1. This is indeed not a regression in compare(). I've checked in 3.10.8, the reason I didn't get this error is that in my previous version a null value returned to a JSONB field would be returned as null, and with 3.12.4 we get a NullNode object instead. Not sure exactly why this happens, I had to make several changes to support the jooq upgrade including in my JsonNodeBinder, I need to investigate this further. 

2. A simple solution that will always satisfy the Comparable/equals contract, is to first check `equals()` in `compare()` and return 0 in that case. This will solve my little edge case and I am happy to provide a pull request of this. However this will not allow any natural ordering of such records of course.

Another easy option is to ignore values that are not Comparable or primitives, when comparing, and document this. Personally I haven't found a real use case for comparing records after they have been fetched, normally I need them in the order defined in the SQL.

Perhaps it makes sense that `Binding` interface will be `Binding<T extends Comparable, U extends Comparable>`? This might be a breaking change but seems quite robust.

And of course the SPI solution makes sense, especially since the Comparable contract can be easily broken when using binding. I think I saw other issues looking for changing how records are compared (for example using PK only).

**Comment by amitgol** [comment]
As a side note, I found the root cause specific to my use case, it is because JSONB object is now wrapping the null value so when previously I checked in my converter `if (dataBaseObject == null)` I now need to check `if (dataBaseObject.data() == null)`

**Comment by lukaseder** [comment]
> 1. This is indeed not a regression in compare(). 

Well, that's good news (for us) :)

> I've checked in 3.10.8, the reason I didn't get this error is that in my previous version a null value returned to a JSONB field would be returned as null, and with 3.12.4 we get a NullNode object instead. Not sure exactly why this happens, I had to make several changes to support the jooq upgrade including in my JsonNodeBinder, I need to investigate this further.

Yes, `json == null` and `json.data() == null` are different things in jOOQ and in SQL. That was a recent fix, see also: https://github.com/jOOQ/jOOQ/issues/9612

> 2\. A simple solution that will always satisfy the Comparable/equals contract, is to first check `equals()` in `compare()` and return 0 in that case

That's not a good solution (e.g. it would be wrong for `BigDecimal`). We definitely don't want to break any existing logic for this edge case. If there exists natural ordering (or we have an SPI to provide ordering) for the type, then we *must* use that.

> This will solve my little edge case and I am happy to provide a pull request of this. 

No need, thank you. The effort of merging this into dual licensed jOOQ is bigger than me just going ahead with a fix. See also: https://github.com/jOOQ/jOOQ/blob/main/CONTRIBUTING.md

> Perhaps it makes sense that `Binding` interface will be `Binding<T extends Comparable, U extends Comparable>`? This might be a breaking change but seems quite robust.

How's the saying? Using a sledgehammer to crack a nut? 

There's absolutely no need for `Binding` (or anything referencing `<T>` or `<U>` types, really) to require these types be `Comparable`, except for this edge case in this rather rare feature of being able to compare records. Perhaps, making `Record` comparable in #2107 was "daring", but spock's pretending `a == b` being the same thing as `a.compareTo(b) == 0` is just plain wrong. According to the `Comparable` contract:

> [Javadoc] It is strongly recommended (though not required) that natural orderings be consistent with equals

Surely, spock just tried to be clever here, and work around the weird behaviour of `BigDecimal::equals` (with `new BigDecimal("1.0")` not being equal to `new BigDecimal("1.00")` but comparing as the same value). I'd consider that a bug in spock (and in `BigDecimal`), but not a justification to break jOOQ's APIs.

I'll go ahead and prevent the `ClassCastException` with the suggested algorithm:

1. If a column type is comparable, use that (keep the status quo)
2. If we have an SPI-provided `Comparator` for the column type, then use that (a setting would allow for switching 1) and 2)'s order). Switching is backwards compatible, because by default, we don't have any comparator providers.
3. In all other cases, return `((Integer) a.hashCode()).compareTo(b.hashCode())`. This will produce random ordering, but at least, it should be consistent with the contracts of `equals()`, `hashCode()`, and `compareTo()`. We don't have to check `equals()` because it is already required to be consistent with `hashCode()`.

I've created a separate issue for the SPI, which will be implemented in a later release: https://github.com/jOOQ/jOOQ/issues/10617

**Comment by lukaseder** [comment]
There's a slight chance of someone relying on this `ClassCastException`, which is why I'm not going to backport the fix to 3.13.5

**Comment by lukaseder** [comment]
Fixed for jOOQ 3.14.0. Thanks again for your report.

**Comment by lukaseder** [comment]
This fix has introduced a regression:

- https://github.com/jOOQ/jOOQ/issues/17580
